src/main/scala/cic.scala

```diff
@@ -2,110 +2,97 @@ package cic
+
+  /* CIC implementation */
+
+  /* sample pdm data on pdm_pulse for first integrator stage */
+  val pdm_bit = RegEnable(Mux(io.pdm.data, 1.S, -1.S), 1.S(width.W), pdm_pulse)
+
+  val itgrOutput = {
+    /* Integrator stage definition */
+    def getItgrStage(inp: SInt): SInt = {
+      val itgr  = RegInit(0.S(width.W))
+      when(pdm_pulse)
+      {
+        itgr := itgr + inp
+      }
+      itgr
     }
 
+    /* integrator stages */
+    val itgr_outputs = Wire(Vec(c.nStages, SInt(width.W)))
 
-    /* Integrator stages */
-    val itgr  = RegInit(VecInit(Seq.fill(N){0.S(width.W)}))
-    when(pdm_edge){
-      itgr.zipWithIndex.foreach {
-        case(itgvalue, 0) =>
-          itgr(0) := itgr(0) + pdm_bit
-        case(itgvalue, i) =>
-          itgr(i) := itgr(i) + itgr(i-1)
-      }
+    itgr_outputs(0) := getItgrStage(pdm_bit)
+    1 until c.nStages foreach {i=>
+      itgr_outputs(i) := getItgrStage(itgr_outputs(i - 1))
     }
-
-    /* Downsampler */
-   val sampleCount = RegInit(R.U)
-   val dec_pulse = RegInit(false.B)
-
-   dec_pulse := false.B
-   when(pdm_edge){
-      when(sampleCount === 0.U){
-        sampleCount := R.U
-        dec_pulse := true.B
-      }.otherwise {
-        sampleCount := sampleCount - 1.U
-      }
+    itgr_outputs.reverse.head
+  }
+
+  io.pcm.bits := {
+    /* Comb stage definition */
+    def getCombStage(inp: SInt): SInt = {
+      val delayed_value = ShiftRegister( inp, c.combDelay, 0.S(width.W), dec_pulse)
+      RegEnable(inp - delayed_value, 0.S(width.W), dec_pulse)
     }
 
-    /* Comb filter stages */
-    val comb_reg = RegInit(VecInit(Seq.fill(N){0.S(width.W)}))
-    val delayed_values = RegInit(VecInit(Seq.fill(N){0.S(width.W)}))
-    comb_reg.zipWithIndex.foreach {
-      case(cbreg, 0) => {
-        delayed_values(0) := ShiftRegister(itgr(N-1), M, dec_pulse)
-        when(dec_pulse) {
-          comb_reg(0) := itgr(N-1) - delayed_values(0)
-        }
-      }
-      case(cbreg, i) => {
-        delayed_values(i) := ShiftRegister(comb_reg(i-1), M, dec_pulse)
-        when(dec_pulse) {
-          comb_reg(i) := comb_reg(i-1) - delayed_values(i)
-        }
-      }
+    /* comb stages */
+    val comb_outputs = Wire(Vec(c.nStages, SInt(width.W)))
+
+    comb_outputs(0) := getCombStage(itgrOutput)
+    1 until c.nStages foreach {i=>
+      comb_outputs(i) := getCombStage(comb_outputs(i - 1))
     }
+    comb_outputs.reverse.head
+  }
 
-    io.pcm.valid := dec_pulse
-    io.pcm.bits := comb_reg(N-1)
+  /* PCM valid */
+  io.pcm.valid := dec_pulse
 }
 
 object CICDriver extends App {
-    println("Some scala tests")
-    val cicpar = CICParam(5, 32, 1)
-    cicpar.display()
-    println("Generate CIC verilog")
-    (new ChiselStage).emitVerilog(new CIC())
+  println("Some scala tests")
+  val c = CICParams(5, 32, 1)
+  print(c)
+  println("Generate CIC verilog")
+  (new ChiselStage).execute(
+    Array("-X", "verilog"),
+    Seq(ChiselGeneratorAnnotation(() => new CIC())))
 }
```

Comment:
why not `(new ChiselStage).emitVerilog(new CIC())` ? It's more compact no ?

Reply:
Martoni, you are right. emitVerilog is a utility method to emit HDL for one module. For multiple modules however, execute method is a preferred way and I am accustomed to using it as a factory in emitter objects. Nevertheless, you can revert back to your old style which is perfectly fine.